test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
+
+	// create multiple clones
+	clonedContainers := []string{}
+	for i := 0; i < nClones; i++ {
+		clonedPodID, clonedContainerID := createClonedContainer(ctx, t, client, templatePodID, templateContainerID, i)
+		// cleanup
+		defer removePodSandbox(t, client, ctx, clonedPodID)
+		defer stopPodSandbox(t, client, ctx, clonedPodID)
+		defer removeContainer(t, client, ctx, clonedContainerID)
+		defer stopContainer(t, client, ctx, clonedContainerID)
+		clonedContainers = append(clonedContainers, clonedContainerID)
+	}
+
+	for i := 0; i < nClones; i++ {
+		verifyContainerExec(ctx, t, client, clonedContainers[i])
+	}
+}
+
+// Test if a normal container can be created inside a clond pod alongside the cloned
+// container.
+// TODO(ambarve): This doesn't work as of now. Enable this test when the bug is fixed.
+func DisabledTest_NormalContainerInClonedPod_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+
+	// create template pod & container
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
+
+	// create a cloned pod and a cloned container
+	cloneSandboxRequest := getClonedPodConfig(1, templatePodID)
+	cloneContainerRequest := getClonedContainerConfig(1, templateContainerID)
+	clonedPodID, clonedContainerID := createPodAndContainer(ctx, t, client, cloneSandboxRequest, cloneContainerRequest)
+	defer removePodSandbox(t, client, ctx, clonedPodID)
+	defer stopPodSandbox(t, client, ctx, clonedPodID)
+	defer removeContainer(t, client, ctx, clonedContainerID)
+	defer stopContainer(t, client, ctx, clonedContainerID)
+
+	// create a normal container in cloned pod
+	stdContainerRequest := getStandardContainerConfig("standard-container")
+	stdContainerRequest.PodSandboxId = clonedPodID
+	stdContainerRequest.SandboxConfig = cloneSandboxRequest.Config
+	stdContainerID := createContainer(t, client, ctx, stdContainerRequest)
+	startContainer(t, client, ctx, stdContainerID)
+	defer removeContainer(t, client, ctx, stdContainerID)
+	defer stopContainer(t, client, ctx, stdContainerID)
+
+	verifyContainerExec(ctx, t, client, clonedContainerID)
+	verifyContainerExec(ctx, t, client, stdContainerID)
+}
+
+// A test for cloning multiple pods first and then cloning one container in each
+// of those pods.
+func Test_CloneContainersWithClonedPodPool_WCOW(t *testing.T) {
```

Comment:
What functionality does this test? I'm a little confused. Is the goal just to show that you can have multiple pod clones at once that function as expected? Would creating two pods with one container each be sufficient to show that?

Reply:
This tests that you can clones `n` pods first and then clone `n` containers into them (which is what cosmos analytics team will do). We can do that with 2 or 3 pod/containers. I felt 3 was better but we can switch to 2 if you want.

Reply:
I see. I think this is fine then.